pym/bob/cmds/build/share.py

```diff
@@ -0,0 +1,296 @@
+                with open(os.path.join(tmpSharedPath, "pkg.json"), "w") as f:
+                    json.dump({
+                        "hash" : asHexStr(sharedHash),
+                        "size" : actualSize,
+                        "users" : [ os.path.abspath(workspace) ]
+                    }, f)
+
+                # Atomic install. Loosing the race is not considered a problem.
+                try:
+                    os.rename(tmpSharedPath, sharedPath)
+                except OSError as e:
+                    if e.errno in (errno.ENOTEMPTY, errno.EEXIST):
+                        return sharedPath, False
+                    raise
+
+                # Add to quota
+                repoSize = self.__addPackage(buildId, actualSize)
+        except OSError as e:
+            raise BuildError("Error installing shared package: " + str(e))
+
+        if (self.__quota is not None) and (repoSize > self.__quota):
+            if self.__autoClean:
+                repoSize = self.gc(False, False, newPkg=sharedPath)
+                if repoSize > self.__quota: warnGcDidNotHelp.show(self.__path)
+            else:
+                warnRepoSize.show(self.__path)
+
+        return sharedPath, True
+
+    def gc(self, pruneUsed, pruneUnused, dryRun=False, progress=lambda x: None, newPkg=None):
+        if (self.__quota is None) and not pruneUnused:
+            return None
+        if not os.path.isdir(self.__path):
+            return 0
+
+        # Create a temporary attic directory. All garbage collected packages
+        # are moved there to delete them without holding any locks.
+        repoSize = 0
+        with tempfile.TemporaryDirectory(dir=self.__path) as attic:
+            # Get exclusive lock on repository. Prohibits further installations
+            # and usage of packages.
+            candidates = []
+            with OpenLocked(os.path.join(self.__path, "repo.json"), "r+", True) as rf:
+                repoMeta = json.load(rf)
+
+                # Scan all packages
+                for pkg, size in repoMeta.get("pkgs", {}).items():
+                    repoSize += size
+                    pkgPath = self.__buildPath(bytes.fromhex(pkg))
+                    try:
+                        with OpenLocked(os.path.join(pkgPath, "pkg.json"), "r", False) as pf:
+                            pkgMeta = json.load(pf)
+                            pkgTime = os.fstat(pf.fileno()).st_mtime_ns
+                            pkgUnused = checkUnused(pkgMeta, pkgPath) and (pkgPath != newPkg)
+                        if pkgUnused or pruneUsed:
+                            candidates.append((pkgUnused, pkgTime, size, pkg))
+                    except FileNotFoundError:
+                        pass
+
+                # Move all candidates to the attic until we are under the quota
+                for pkgUnused, _, pkgSize, pkgBuildId in sorted(candidates):
```

Comment:
The manpage states that packages are deleted *from oldest usage to newest*, now you're ignoring `pkgTime`. Is this intentional? Is the behaviour ensured implicitly?

Reply:
This is taken care of by the `sorted()` on the right side. It sorts the tuples by comparing the elements. So the unused packages go first, then they are sorted by time (and then size and buildid but who cares ;))...

Reply:
Of course it does :exploding_head: